orm/model/ormtable/build.go

```diff
@@ -0,0 +1,295 @@
+
+	// SingletonDescriptor is an optional singleton descriptor to be explicitly used.
+	// Generally this should be nil and the table descriptor
+	// should be pulled from the singleton message option. SingletonDescriptor
+	// cannot be used together with TableDescriptor.
+	SingletonDescriptor *ormv1alpha1.SingletonDescriptor
+
+	// TypeResolver is an optional type resolver to be used when unmarshaling
+	// protobuf messages.
+	TypeResolver TypeResolver
+
+	// JSONValidator is an optional validator that can be used for validating
+	// messaging when using ValidateJSON. If it is nil, DefaultJSONValidator
+	// will be used
+	JSONValidator func(proto.Message) error
+
+	// GetBackend is an optional function which retrieves a Backend from the context.
+	// If it is nil, the default behavior will be to attempt to retrieve a
+	// backend using the method that WrapContextDefault uses. This method
+	// can be used to imlement things like "store keys" which would allow a
+	// table to only be used with a specific backend and to hide direct
+	// access to the backend other than through the table interface.
+	GetBackend func(context.Context) (Backend, error)
+
+	// GetReadBackend is an optional function which retrieves a ReadBackend from the context.
+	// If it is nil, the default behavior will be to attempt to retrieve a
+	// backend using the method that WrapContextDefault uses.
+	GetReadBackend func(context.Context) (ReadBackend, error)
+}
+
+// TypeResolver is an interface that can be used for the protoreflect.UnmarshalOptions.Resolver option.
+type TypeResolver interface {
+	protoregistry.MessageTypeResolver
+	protoregistry.ExtensionTypeResolver
+}
+
+// Build builds a Table instance from the provided Options.
+func Build(options Options) (Table, error) {
+	messageDescriptor := options.MessageType.Descriptor()
+
+	getReadBackend := options.GetReadBackend
+	if getReadBackend == nil {
+		getReadBackend = getReadBackendDefault
+	}
+	getBackend := options.GetBackend
+	if getBackend == nil {
+		getBackend = getBackendDefault
+	}
+
+	table := &tableImpl{
+		primaryKeyIndex: &primaryKeyIndex{
+			indexers:       []indexer{},
+			getBackend:     getBackend,
+			getReadBackend: getReadBackend,
+		},
+		indexes:               []Index{},
+		indexesByFields:       map[fieldNames]concreteIndex{},
+		uniqueIndexesByFields: map[fieldNames]UniqueIndex{},
+		entryCodecsById:       map[uint32]ormkv.EntryCodec{},
+		typeResolver:          options.TypeResolver,
+		customJSONValidator:   options.JSONValidator,
+	}
+
+	pkIndex := table.primaryKeyIndex
+
+	tableDesc := options.TableDescriptor
+	if tableDesc == nil {
+		tableDesc = proto.GetExtension(messageDescriptor.Options(), ormv1alpha1.E_Table).(*ormv1alpha1.TableDescriptor)
+	}
+
+	singletonDesc := options.SingletonDescriptor
+	if singletonDesc == nil {
+		singletonDesc = proto.GetExtension(messageDescriptor.Options(), ormv1alpha1.E_Singleton).(*ormv1alpha1.SingletonDescriptor)
+	}
+
+	if tableDesc != nil {
```

Comment:
this if,elseif,else can be covered by a switch statement to make it more readable

Reply:
done